File: apps/mybookkeeper/backend/app/schemas/rent/rent_ledger_response.py
"""The full rent picture for one tenant."""
from __future__ import annotations

import datetime as _dt
import uuid
from decimal import Decimal

from pydantic import BaseModel

from app.schemas.rent.rent_charge_response import RentChargeResponse
from app.schemas.rent.rent_payment_response import RentPaymentResponse
from app.schemas.rent.rent_period_summary import RentPeriodSummary
from app.schemas.rent.rent_schedule_response import RentScheduleResponse


class RentLedgerResponse(BaseModel):
    applicant_id: uuid.UUID
    as_of: _dt.date

    schedules: list[RentScheduleResponse]
    charges: list[RentChargeResponse]
    payments: list[RentPaymentResponse]

    # The period containing ``as_of``. NULL when no schedule covers today —
    # either none exists yet or the tenancy has ended.
    current_period: RentPeriodSummary | None = None

    total_charged: Decimal
    total_paid: Decimal
    # Positive = tenant owes. Negative = tenant is in credit (paid ahead).
    balance: Decimal
    # Payments received that no charge has consumed yet.
    unapplied_credit: Decimal

File: apps/mybookkeeper/frontend/src/shared/types/rent/rent-ledger-response.ts
import type { RentCharge } from "./rent-charge";
import type { RentPayment } from "./rent-payment";
import type { RentPeriodSummary } from "./rent-period-summary";
import type { RentSchedule } from "./rent-schedule";

/** The full rent picture for one tenant. */
export interface RentLedgerResponse {
  applicant_id: string;
  as_of: string;
  schedules: RentSchedule[];
  charges: RentCharge[];
  payments: RentPayment[];
  /** The period containing `as_of`, or null when no schedule covers it. */
  current_period: RentPeriodSummary | null;
  total_charged: string;
  total_paid: string;
  /** Positive = tenant owes. Negative = tenant is in credit. */
  balance: string;
  unapplied_credit: string;
}
